### ISSUE TITLE
I/O Error reading directory

### ISSUE THREAD

**Issue by daFritz84** [EnzephaloN/ncpfs_dkms#3]
First of all: Thank you for maintaining this project in 2021! While Novell stuff has fallen into obscurity for most people, there are still a chosen few (like myself) who need to interface with such monstrosities even today.

---

I use ncpfs mounting a share via fstab

```
# ncp netware mounts
OBFUSCATED            /mnt/OES/DEVEL          ncp    noatime,timeout=600,ipserver=obfuscated.com,passwd=private,volume=DEVEL,uid=1001,gid=100    0 0
```

and from time to time the connection gets lost (presumably), i.e. if I do a directory listing I see an io error:

```
chef@microfocus-netware-bridge:/mnt/OES/DEVEL> ls -sla
ls: reading directory '.': Input/output error
total 0
```

Right now I need to unmount and remount the directory to get stuff working again. I can also see that the log file of the system is cluttered with warnings (e.g.):

```
2021-05-31T11:07:12.217306+02:00 microfocus-netware-bridge kernel: [2657772.271004] ------------[ cut here ]------------
2021-05-31T11:07:12.217307+02:00 microfocus-netware-bridge kernel: [2657772.271017] WARNING: CPU: 0 PID: 19549 at ../lib/iov_iter.c:1135 iov_iter_kvec+0x51/0x70
2021-05-31T11:07:12.217320+02:00 microfocus-netware-bridge kernel: [2657772.271018] Modules linked in: nls_utf8 af_packet xt_tcpudp ip6t_rpfilter ip6t_REJECT nf_reject_ipv6 ipt_REJECT nf_reject_ipv4 xt_conntrack ip_set nfnetlink ebtable_nat ebtable_broute ip6table_nat ip6table_mangle ip6table_raw ip6table_security iptable_nat nf_nat nf_conntrack nf_defrag_ipv6 nf_defrag_ipv4 iptable_mangle iptable_raw iptable_security ebtable_filter ebtables iscsi_ibft iscsi_boot_sysfs ip6table_filter ip6_tables rfkill iptable_filter ip_tables x_tables bpfilter intel_rapl_msr intel_rapl_common crct10dif_pclmul crc32_pclmul ghash_clmulni_intel iTCO_wdt iTCO_vendor_support aesni_intel aes_x86_64 crypto_simd cryptd glue_helper pcspkr i2c_i801 lpc_ich virtio_net net_failover failover virtio_balloon joydev button ncpfs(OE) btrfs libcrc32c xor hid_generic usbhid raid6_pq qxl drm_kms_helper syscopyarea sysfillrect sysimgblt fb_sys_fops crc32c_intel xhci_pci ttm xhci_hcd serio_raw sr_mod cdrom usbcore drm virtio_rng virtio_scsi qemu_fw_cfg sg dm_multipath dm_mod scsi_dh_rdac
2021-05-31T11:07:12.217321+02:00 microfocus-netware-bridge kernel: [2657772.271037]  scsi_dh_emc scsi_dh_alua
2021-05-31T11:07:12.217321+02:00 microfocus-netware-bridge kernel: [2657772.271039] CPU: 0 PID: 19549 Comm: ls Tainted: G        W  OE      5.3.18-lp152.72-preempt #1 openSUSE Leap 15.2
2021-05-31T11:07:12.217322+02:00 microfocus-netware-bridge kernel: [2657772.271040] Hardware name: oVirt RHEL, BIOS 1.14.0-1.el8 04/01/2014
2021-05-31T11:07:12.217322+02:00 microfocus-netware-bridge kernel: [2657772.271041] RIP: 0010:iov_iter_kvec+0x51/0x70
2021-05-31T11:07:12.217323+02:00 microfocus-netware-bridge kernel: [2657772.271042] Code: 89 43 10 89 33 48 83 c4 20 5b c3 48 c7 c7 a0 e1 2c a0 4c 89 44 24 18 48 89 4c 24 10 48 89 54 24 08 89 74 24 04 e8 1d 91 c2 ff <0f> 0b 4c 8b 44 24 18 48 8b 4c 24 10 48 8b 54 24 08 8b 74 24 04 eb
2021-05-31T11:07:12.217381+02:00 microfocus-netware-bridge kernel: [2657772.271648] RSP: 0018:ffffb8d180747828 EFLAGS: 00010282
2021-05-31T11:07:12.217382+02:00 microfocus-netware-bridge kernel: [2657772.271649] RAX: 0000000000000024 RBX: ffff8e90c5efd548 RCX: 0000000000000006
2021-05-31T11:07:12.217382+02:00 microfocus-netware-bridge kernel: [2657772.271650] RDX: 0000000000000000 RSI: 0000000000000002 RDI: ffff8e913d819a10
2021-05-31T11:07:12.217383+02:00 microfocus-netware-bridge kernel: [2657772.271650] RBP: ffff8e90c5efd500 R08: 000b4dfbb499b0d1 R09: 0000000000000024
2021-05-31T11:07:12.217383+02:00 microfocus-netware-bridge kernel: [2657772.271650] R10: ffff8e9136b63020 R11: 00000000000b0940 R12: 000000000000001b
2021-05-31T11:07:12.217384+02:00 microfocus-netware-bridge kernel: [2657772.271651] R13: ffff8e90c5efd510 R14: ffffb8d180769000 R15: ffff8e90c5efd500
2021-05-31T11:07:12.217384+02:00 microfocus-netware-bridge kernel: [2657772.271652] FS:  00007f03128c3e80(0000) GS:ffff8e913d800000(0000) knlGS:0000000000000000
2021-05-31T11:07:12.217385+02:00 microfocus-netware-bridge kernel: [2657772.271652] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
2021-05-31T11:07:12.217385+02:00 microfocus-netware-bridge kernel: [2657772.271653] CR2: 0000564cdb021064 CR3: 000000003aca6005 CR4: 00000000001606f0
2021-05-31T11:07:12.217386+02:00 microfocus-netware-bridge kernel: [2657772.271655] Call Trace:
2021-05-31T11:07:12.217386+02:00 microfocus-netware-bridge kernel: [2657772.271658]  ncpdgram_start_request+0x74/0x100 [ncpfs]
2021-05-31T11:07:12.217332+02:00 microfocus-netware-bridge kernel: [2657772.271066]  ? schedule_hrtimeout_range_clock+0x99/0x100
2021-05-31T11:07:12.217333+02:00 microfocus-netware-bridge kernel: [2657772.271067]  ? remove_wait_queue+0x12/0x50
2021-05-31T11:07:12.217333+02:00 microfocus-netware-bridge kernel: [2657772.271068]  ? copyout+0x22/0x30
2021-05-31T11:07:12.217334+02:00 microfocus-netware-bridge kernel: [2657772.271083]  ? btrfs_verify_level_key+0xba/0x170 [btrfs]
2021-05-31T11:07:12.217334+02:00 microfocus-netware-bridge kernel: [2657772.271098]  ? btrfs_get_token_64+0x103/0x120 [btrfs]
2021-05-31T11:07:12.217335+02:00 microfocus-netware-bridge kernel: [2657772.271108]  ? btrfs_show_options+0x3b4/0x530 [btrfs]
2021-05-31T11:07:12.217335+02:00 microfocus-netware-bridge kernel: [2657772.271109]  ? ncp_unlink+0x1d0/0x1d0 [ncpfs]
2021-05-31T11:07:12.217336+02:00 microfocus-netware-bridge kernel: [2657772.271110]  ? __d_lookup_rcu+0x12e/0x190
2021-05-31T11:07:12.217349+02:00 microfocus-netware-bridge kernel: [2657772.271112]  ? lookup_fast+0x2d4/0x2f0
2021-05-31T11:07:12.217350+02:00 microfocus-netware-bridge kernel: [2657772.271114]  ? walk_component+0x48/0x300
2021-05-31T11:07:12.217350+02:00 microfocus-netware-bridge kernel: [2657772.271115]  ? path_lookupat+0x6e/0x210
2021-05-31T11:07:12.217350+02:00 microfocus-netware-bridge kernel: [2657772.271117]  ? filename_lookup+0xb6/0x190
2021-05-31T11:07:12.217351+02:00 microfocus-netware-bridge kernel: [2657772.271119]  ? __check_object_size+0xd4/0x1a0
2021-05-31T11:07:12.217351+02:00 microfocus-netware-bridge kernel: [2657772.271120]  ? strncpy_from_user+0x47/0x180
2021-05-31T11:07:12.217352+02:00 microfocus-netware-bridge kernel: [2657772.271121]  ? getname_flags+0x66/0x1d0
2021-05-31T11:07:12.217352+02:00 microfocus-netware-bridge kernel: [2657772.271123]  ? vfs_statx+0x73/0xe0
2021-05-31T11:07:12.217353+02:00 microfocus-netware-bridge kernel: [2657772.271124]  ? vfs_statx+0x73/0xe0
2021-05-31T11:07:12.217354+02:00 microfocus-netware-bridge kernel: [2657772.271126]  ? __do_sys_newlstat+0x39/0x70
2021-05-31T11:07:12.217354+02:00 microfocus-netware-bridge kernel: [2657772.271128]  ? do_syscall_64+0x65/0x1f0
2021-05-31T11:07:12.217355+02:00 microfocus-netware-bridge kernel: [2657772.271129]  ? entry_SYSCALL_64_after_hwframe+0x44/0xa9
2021-05-31T11:07:12.217375+02:00 microfocus-netware-bridge kernel: [2657772.271131] ---[ end trace b6b85f8d24f0ab8a ]---
2021-05-31T11:07:12.217376+02:00 microfocus-netware-bridge kernel: [2657772.271610] ------------[ cut here ]------------
```

Do you have any tip/suggestion what I might try to further investigate this issue or get rid of the warnings to improve the message log?

PS: Anyways I owe you a coffee/beer/tea or whatever beverage you might prefer. So if you have some kind of donation site going on (i.e. Patreon or the like) I will drop some coins. :)

**Comment by EnzephaloN** [comment]
Hello @daFritz84 .
Thanks for your kudos, but these have to go to my teammate which does the code - I will tell him about ;-) . I'm just pushing the code into github.

I asked around and none of my teammates have errors like you, so we can't provide any tips/suggestions to you. Sorry.
Does the error occur, if the server (share) is absent for some time?

But we recognized, that the module is broken for kernel > 5.12 - which is sad. I hope my teammate will find a way.

best regards
EnzephaloN

**Comment by daFritz84** [comment]
Thanks for asking around,

I'm using OpenSuse Leap 15.2 with Kernel 5.3.18. I figured this will give me the greatest chance of success, since Suse was once owned by Novell.

```
uname -a
Linux microfocus-netware-bridge 5.3.18-lp152.72-preempt #1 SMP PREEMPT Wed Apr 14 10:13:15 UTC 2021 (013936d) x86_64 x86_64 x86_64 GNU/Linux
```

If I may ask: Which Linux Distro's do you use in conjunction with ncpfs-dkms?

**Comment by daFritz84** [comment]
the function iov_iter_kvec in Kernel 5.3.18 looks like this:
```
void iov_iter_kvec(struct iov_iter *i, unsigned int direction,
			const struct kvec *kvec, unsigned long nr_segs,
			size_t count)
{
	WARN_ON(direction & ~(READ | WRITE));
	i->type = ITER_KVEC | (direction & (READ | WRITE));
	i->kvec = kvec;
	i->nr_segs = nr_segs;
	i->iov_offset = 0;
	i->count = count;
}
```
So the messages in the log files are in fact the WARN_ON macro complaining that the direction is neither WRITE nor READ.

The offending call to iov_iter_kvec in function ncpdgram_start_request (sock.c) looks like this:
```
iov_iter_kvec(&req->from, ITER_KVEC, req->tx_iov + 1, signlen ? 2 : 1, len + signlen);
```

What looks a little bit odd to me is the ITER_KVEC constant. In fact similar calls in the same source file use either READ or WRITE as the second parameter to this call.

TLDR: The ITER_KVEC does not look right to me? Can you confirm this is correct? 
I also could not trace the source of this constant so far. However it does compile, so it must be defined somewhere.

**Comment by EnzephaloN** [comment]
Hello

My teammates and I are on multiple distros, some use Gentoo, some Suse or PcLinux, some like me use Debian.

We didn't wrote all that code, we just patched the code for working on newer kernels.

At this time we try to patch the code for working on kernel > 5.12 - which is a lot of work because of API-changes.

**Comment by EnzephaloN** [comment]
> TLDR: The ITER_KVEC does not look right to me? Can you confirm this is correct?
> I also could not trace the source of this constant so far. However it does compile, so it must be defined somewhere.

as we patched to kernel 4.20, we changed parameters of all iov_iter_kvec calls in sock.c from "WRITE | ITER_KVEC" or "READ | ITER_KVEC" to "WRITE" or "READ". We missed or failed a change in ncpdgram_start_request. I will fix this. Can you please confirm if this fix works for you?

**Comment by daFritz84** [comment]
Finally got around to test the changes.

The latest tip of the master branch fails for me due to the changes in 918f1b6c641a6511993dae133cd4162840c4f56c. But this is probbaly to be expected. 

Anyways. I tested 918f1b6c641a6511993dae133cd4162840c4f56c and it got rid of the warnings in the message log. 
Can you backport this fix to the older kernel releases please 😊 ?

I see you did some changes to the branching structure. The last time I had a look, the master branch was qualified for Kernels > 5.2. Now there is a dedicated branch for 5.6, and so on ... So I assume, I need to switch to the Kernel 4.20 branch, given that I'm running 5.3.18.

**Comment by EnzephaloN** [comment]
I did a cleanup and fixed the missed change in all branches where it was needed.
Accidently I removed kernel_5.2-branch. I fixed this too.
If you are on 5.3.18, please use the kernel_5.2 branch.

**Comment by daFritz84** [comment]
Works like a charm. Thanks for all your help! 🙇 

I will monitor the system for more 'I/O errors', but close the issue for now. With the warnings gone I should have a fair chance of getting getting to the culprit of this.
